Sources/src/CosmicCORSIKA.cc

```diff
@@ -23,19 +23,74 @@ namespace mu2e {
         _targetBoxYmax(conf.targetBoxYmax()),  // mm
         _targetBoxZmin(conf.targetBoxZmin()), // mm
         _targetBoxZmax(conf.targetBoxZmax()),  // mm
-        _resample(conf.resample()),
-        _compact(conf.compact()),
         _engine(seed),
         _randFlatX(_engine, -(_targetBoxXmax-_targetBoxXmin+_showerAreaExtension)/2, +(_targetBoxXmax-_targetBoxXmin+_showerAreaExtension)/2),
         _randFlatZ(_engine, -(_targetBoxZmax-_targetBoxZmin+_showerAreaExtension)/2, +(_targetBoxZmax-_targetBoxZmin+_showerAreaExtension)/2)
   {
   }
 
-  void CosmicCORSIKA::openFile(FILE *f) {
-    in = f;
-    fread(&_garbage, 4, 1, in);
-  }
+  void CosmicCORSIKA::openFile(ifstream *f, unsigned &runNumber, float &lowE, float &highE)
+  {
+    input = f;
+    _current_event_number = -1;
+    _event_count = 0;
+    _run_number = -1;
+    _infmt = Format::UNDEFINED;
+
+    while( input->read(_buf.ch, 4)) {
+      unsigned reclen = _buf.in[0];
+      // CORSIKA records are in units of 4 bytes
+      if(reclen % 4) {
+        throw std::runtime_error("Error: record size not a multiple of 4");
+      }
+
+      // We will be looking at at least 8 bytes to determine the
+      // input file format, and all real CORSIKA records are longer
+      // than that.
+      if(reclen < 2*4) {
+        throw std::runtime_error("Error: reclen too small");
+      }
+
```

Comment:
I suggest to also add a protection agains reclen > 4*_fbsize_words,  so we do not overwrite memory on corrupt inputs.

Reply:
Added, thanks.